### DISCUSSION TITLE
Web Server Run: Operation could destabilize the runtime

### DISCUSSION THREAD

**Discussion by err0x1a** [SixLabors/ImageSharp#1569] [Q&A] [answered]
Hello, I'm trying to run the library on a Web server, but the error Operation could destabilize the runtime is occurring.

1. This application is configured as Medium Trust.
2. I think the NET Framework is 4.0, I'm talking to the host to update to 4.5.2.
3. OS Windows Server 2008 R2

But would that cause this error?

Attached print of the error stack.
Thanks in advance for any help. 

![Captura de Tela 2021-03-04 às 11 56 55](https://user-images.githubusercontent.com/324835/109983805-eff18200-7ce1-11eb-8123-8fb5a3227662.png)

**Comment by antonfirsov** [comment, marked as answer]
It looks like .NET Framework's CAS considers a particular operation in our code unreliable, and refuses to JIT our code. There is nothing we can do about this, I suggest to relax your CAS settings. (Not an expert on that field, but hopefully there is a way to scope it to your resizer code.)

**Reply by JimBobSquarePants** [reply]
>NET Framework is 4.0, I'm talking to the host to update to 4.5.2

Those are two very non-supported target frameworks.  

@lvisb Our supported target frameworks. [NetStandard 1.3+](https://github.com/dotnet/standard/blob/master/docs/versions/netstandard1.3.md) are listed in our readme, NuGet, and documentation. This means that the absolute minimum .NET Framework version we support is .NET 4.6 though Microsoft recommend using no lower that .NET 4.7.2 to prevent binding redirect issues.

https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support

**Reply by err0x1a** [reply]
Oh, right! Sorry about that! I came from the ImageProcessor and didn't checked out the supported frameworks. I'll try to update to 4.6 or 4.7.2. Thanks!

**Reply by antonfirsov** [reply]
Binding problems may cause this issue too probably? (Not sure)

**Reply by JimBobSquarePants** [reply]
No worries, netstandard is tough for the uninitiated

**Reply by saucecontrol** [reply]
Any amount of  unsafe code (whether using the keyword or not 😄) requires Full Trust.  `peverify.exe` from the .NET Framework SDK will tell you what specifically will fail in a constrained environment.  Needless to say, ImageSharp has quite a few unverifiable things in it:

```
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Processing.Processors.Transforms.ResizeKernel::.ctor][offset 0x00000008] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Processing.Processors.Transforms.ResizeKernel::AlterLeftValue][offset 0x0000000D][found unmanaged pointer][expected unmanaged pointer] Unexpected type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Processing.Processors.Dithering.ErrorDither::Equals][offset 0x0000000F] Cannot change initonly field outside its .ctor.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Processing.Processors.Dithering.OrderedDither::Equals][offset 0x00000001] Cannot change initonly field outside its .ctor.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteBytes][offset 0x00000013] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteBytes][offset 0x00000027] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteBytesDirect][offset 0x0000000A] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteUInt16][offset 0x00000003][found address of Short] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteInt16][offset 0x00000003][found address of Short] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteUInt32][offset 0x00000003][found address of Int32] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteInt32][offset 0x00000003][found address of Int32] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteUInt64][offset 0x00000003][found address of Long] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteInt64][offset 0x00000003][found address of Long] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteSingle][offset 0x00000003][found address of Single] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Metadata.Profiles.Icc.IccDataWriter::WriteDouble][offset 0x00000003][found address of Double] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Formats.Jpeg.Components.Block8x8::GetScalarAt][offset 0x00000000] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Formats.Jpeg.Components.Block8x8::SetScalarAt][offset 0x00000000] Unmanaged pointers are not a verifiable type.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Formats.Jpeg.Components.Encoder.RgbToYCbCrTables::Create][offset 0x0000001F][found address of Int32] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.Formats.Jpeg.Components.Decoder.HuffmanScanBuffer::DecodeHuffman][offset 0x0000001A][found address of Byte] Expected numeric type on the stack.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.ColorSpaces.CieLab::Equals][offset 0x00000043] Cannot change initonly field outside its .ctor.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.ColorSpaces.CieLuv::Equals][offset 0x00000043] Cannot change initonly field outside its .ctor.
[IL]: Error: [SixLabors.ImageSharp.dll : SixLabors.ImageSharp.ColorSpaces.HunterLab::Equals][offset 0x00000043] Cannot change initonly field outside its .ctor.
```

If you GAC an assembly, it will inherit Full Trust, which is why the Framework itself is allowed to have unsafe/unverifiable code and still run in a constrained environment.  That could be a workaround if Medium Trust is an actual requirement.

**Reply by JimBobSquarePants** [reply]
Medium Trust is not even supported by MS anymore https://stackoverflow.com/a/17218344

**Reply by err0x1a** [reply]
The host could upgrade to .NET Framework 4.8, and Full Trust, but still got that error. 

Even with Framework 4.8 I had to copy some DLLs:
- System.Runtime.CompilerServices.Unsafe.dll
- System.Memory.dll
- System.Buffers.dll

This is a common procedure? Or my environment is missing something?

**Reply by JimBobSquarePants** [reply]
Hmmm... I'm not sure why those dependencies are not copied on publish. I'm assuming your application build locally?

Here's the VS dialog in a .NET 4.8 an empty console application.

![image](https://user-images.githubusercontent.com/385879/111016339-29428580-83a5-11eb-93fa-318c49556c0e.png)

Here's my packages config after installing the library into an empty console application.

```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="SixLabors.ImageSharp" version="1.0.3" targetFramework="net48" />
  <package id="System.Buffers" version="4.5.1" targetFramework="net48" />
  <package id="System.Memory" version="4.5.4" targetFramework="net48" />
  <package id="System.Numerics.Vectors" version="4.5.0" targetFramework="net48" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.3" targetFramework="net48" />
</packages>
```

Here's the contents of the bin folder following a build.
![image](https://user-images.githubusercontent.com/385879/111016473-23996f80-83a6-11eb-8a6f-5d3dc45f27b2.png)

As you can see everything is present and correct.

**Reply by err0x1a** [reply]
I am not very familiar with Nuget, but several versions of each DLL have been downloaded, so there is a folder for NET Standard, another folder for Framework 4.X, anyway, I will try to install these versions as in your project and I will try again. I believe I just didn't upload System.Numerics.Vectors to the server.

Anyway I am not very confident that it will work, I'm about to give up, but this has nothing to do with the library. I always have these problems with ASP.NET, it is always a headache to install some libraries that require more resources.

**Reply by JimBobSquarePants** [reply]
How are you publishing to your server? You should never have to [manually copy files](https://docs.microsoft.com/en-us/visualstudio/deployment/quickstart-deploy-to-a-web-site?view=vs-2019#publish-to-a-web-site).

**Reply by antonfirsov** [reply]
> so there is a folder for NET Standard, another folder for Framework 4.X

Normally you don't need to care about internal folder structure of NuGet. Make sure you don't reference the dll-s manually, but use use Visual Studio's NuGet tools to add package references to your project.
